### PR TITLE
[WFCORE-4361] Add evidence decoders in the Elytron subsystem

### PR DESCRIPTION
**Pull request by fjuma**
https://issues.jboss.org/browse/WFCORE-4361
https://issues.jboss.org/browse/EAP7-1173

This also contains the Elytron subsystem version bump included in #3777

To pass CI, this will require an Elytron upgrade but this is ready for the subsystem changes to be reviewed.

Analysis document: https://github.com/wildfly/wildfly-proposals/pull/203
Capabilities PR: https://github.com/wildfly/wildfly-capabilities/pull/34
WildFly PR: https://github.com/wildfly/wildfly/pull/12352
Depends on https://github.com/wildfly-security/wildfly-elytron/pull/1294

### CODE REVIEW

**Comment by jmesnil** [comment]
@fjuma I have merged #3777 (electron subsystem bump) and #3821 (elytron component upgrades), could you please remove the subsystem bump commits from this PR please?

**Comment by darranl** [comment]
@jmesnil This is a GitHub bug not an issue with this PR, I will have a look at the current PRs.

File: elytron/src/main/java/org/wildfly/extension/elytron/EvidenceDecoderDefinitions.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        }
+    }
+
+    static final SimpleAttributeDefinition ALT_NAME_TYPE = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.ALT_NAME_TYPE, ModelType.STRING, false)
+            .setValidator(new StringAllowedValuesValidator(RFC_822_NAME, DNS_NAME, DIRECTORY_NAME, URI_NAME, IP_ADDRESS, REGISTERED_ID))
+            .setAllowExpression(true)
+            .setRestartAllServices()
+            .build();
+
+    static final SimpleAttributeDefinition SEGMENT = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.SEGMENT, ModelType.INT, true)
```
Comment (OndrejKotek) [inline elytron/src/main/java/org/wildfly/extension/elytron/EvidenceDecoderDefinitions.java]: It could be checked whether the value is non-negative (>=0).

Comment (fjuma) [reply elytron/src/main/java/org/wildfly/extension/elytron/EvidenceDecoderDefinitions.java]: Thanks @OndrejKotek, fixed.

**Comment by darranl** [comment]
@fjuma you also have a conflict but may be worth waiting until after the 1.10.0.CR2 tag as we have another imminent merge likely to conflict with this PR.

**Comment by fjuma** [comment]
I've rebased this one and fixed the conflict.

**Comment by darranl** [comment]
@fjuma With this capability I am just wondering should there be a custom-evidence-decoder resource like we have for the other capabilities?

**Comment by fjuma** [comment]
@darranl This is certainly something that could be added. Currently, the only type of evidence that we have that has a principal associated with it is is ```X509PeerCertificateChainEvidence```. The evidence decoders added in this PR now cover the identifying information that can be extracted from an X.509 certificate (the subject name and subject alternative names). However, custom evidence decoders could be useful for custom evidence types. I can add this in.

**Comment by fjuma** [comment]
@darranl I've updated this PR to add support for a ```custom-evidence-decoder```. I've also updated the analysis document and documentation PR accordingly as well. 

@OndrejKotek ^ FYI. This is a small change.

**Comment by darranl** [comment]
Follow up PR available here https://github.com/wildfly/wildfly-core/pull/3845 addressing merge conflict.

**Pull request closed without merging**